packages/zoe/src/zoe.js

```diff
@@ -230,43 +234,63 @@ const makeZoe = (additionalEndowments = {}) => {
      * other information, such as the terms used in the instance.
      * @param  {object} installationHandle - the unique handle for the
      * installation
+     * @param  {object} roles - an object mapping roleName key to
+     * issuer values
      * @param  {object} terms - arguments to the contract. These
-     * arguments depend on the contract, apart from the `issuers`
-     * property, which is required.
+     * arguments depend on the contract.
      */
-    makeInstance: (installationHandle, userProvidedTerms) => {
+    makeInstance: (installationHandle, roles, terms = harden({})) => {
       const { installation } = installationTable.get(installationHandle);
       const instanceHandle = harden({});
       const contractFacet = makeContractFacet(instanceHandle);
 
-      const makeContractInstance = issuerRecords => {
-        const terms = {
-          ...userProvidedTerms,
-          issuers: issuerRecords.map(record => record.issuer),
-        };
+      const firstCapASCII = /^[A-Z][a-zA-Z0-9_$]*$/;
 
-        const instanceRecord = harden({
-          installationHandle,
-          publicAPI: undefined,
-          issuers: terms.issuers,
-          terms,
-        });
+      // `getOwnPropertyNames` returns all the non-symbol properties
+      // (both enumerable and non-enumerable).
+      const roleNames = Object.getOwnPropertyNames(roles);
+      // We sort to get a deterministic order that is not based on key
+      // insertion order or object creation order.
+      roleNames.sort();
+      harden(roleNames);
 
-        instanceTable.create(instanceRecord, instanceHandle);
-        return Promise.resolve(
-          installation.makeContract(contractFacet, terms),
-        ).then(value => {
-          const { invite, publicAPI } = value;
-          instanceTable.update(instanceHandle, { publicAPI });
-          return invite;
-        });
-      };
+      // Insist that there are no symbol properties.
+      assert(
+        Object.getOwnPropertySymbols(roles).length === 0,
+        details`no symbol properties allowed`,
+      );
+
+      // Insist all key characters are ascii and keys start with a
+      // capital letter.
+      assert(
+        roleNames.every(name => firstCapASCII.test(name)),
```

Comment:
Move the assert inside the loop. Assert for each roleName separately, so the diagnostic includes the offending roleName.